pmdarima/arima/_context.py

```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+#
+# re-entrant, reusable context manager to store execution context
+#
+
+import threading
+
+# thread local value to store the context info
+_ctx = threading.local()
```

Comment:
I have a few concerns about using a single global thread local to store all the attributes:

1. A nested context would break (I don't know who would ever do this, but if they did we'd have a problem..):

  ```python
  with StepwiseContext(max_steps=5, max_dur=30) as ctx1:
      with StepwiseContext(max_steps=10, max_dur=15) as ctx2:
          # I would expect ctx1.max_steps to be 5, but I think it would be 10 by the time auto_arima was called
  ```

2. The use of multiple `StepwiseContext`s in a script could produce confusing results:

  ```python
  with StepwiseContext(max_dur=30) as ctx1:
      fit = auto_arima(y, **kwargs)

  with StepwiseContext(max_steps=5) as ctx2:
      # now I think ctx2 would confusingly inherit the max_dur from the first one
  ```

I understand that we need a global so that the stepwise wrapper can acquire it without it being passed... what if rather than a single thread local, the global were a `collections.deque` like:

```python
# _context.py
import collections

_ctx = collections.deque()

class _context():
    def __init__(self, **kwargs):
        for key in kwargs:
            setattr(self, key, kwargs[key])
        _ctx.append(self)  # this can be "popped" via first-in-last-out
```

.. so in the stepwise fit wrapper the context was called more like:

  ```python
  # _auto_solvers.py
  from ._context import _ctx

  ...
  # inside __init__
  try:
      self.exec_context = _ctx.pop()
  except IndexError:
      self.exec_context = None  # not in a context manager
  ```

Thoughts?

Reply:
I think that's a great idea!

Perhaps we may want to leave it as a thread specific instance, but store context instances in a `deque`. An context is added to deque on context manager `__enter__` and removed on `__exit__`, so we can support different usage patterns using `ExitStack`, etc.

To generalize the approach to support other context managers in future, we can wrap access to the thread local into a set of convenience methods (with other supporting fixtures) to something like...

```python
import threading
from abc import ABC, abstractmethod
from enum import Enum
import collections

# thread local value to store the context info
_ctx = threading.local()
_ctx.store = {}

__all__ = ['AbstractContext', 'ContextStore', 'ContextType']


class ContextType(Enum):
    """Context Type Enumeration

    An enumeration of Context Types known to ``ContextStore``

    """
    EMPTY = 0
    STEPWISE = 1


class AbstractContext(ABC):
    """An abstract context manager to store execution context.

    A generic, re-entrant, reusable context manager to store
    execution context in a threading.local instance. Has helper
    methods to iterate over the context info and provide a
    string representation of the context info.

    """
    def __init__(self, **kwargs):
        self.props = kwargs if kwargs else {}

    def __enter__(self):
        ContextStore._add_context(self)

    def __exit__(self, exc_type, exc_val, exc_tb):
        ContextStore._remove_context(self)

    def __getattr__(self, item):
        return self.props[item] if item in self.props else None

    def __contains__(self, item):
        return item in self.props

    def __getitem__(self, item):
        return self.props[item] if item in self.props else None

    def __iter__(self):
        for key in self.props:
            yield key

    def __repr__(self):
        return self.props.__repr__()

    @abstractmethod
    def get_type(self):
        return None


class _emptyContext(AbstractContext):

    def __init__(self):
        super(_emptyContext, self).__init__()

    def get_type(self):
        return ContextType.EMPTY

class ContextStore:
    """A class to wrap access to threading.local() context store

    This class hosts static methods to wrap access to and
    encapsulate the threading.local() instance
    """
    @staticmethod
    def get_context(context_type):
        """Returns most recently added instance of given Context Type

        :param context_type: Context Type to retrieve from the store
        :return: an instance of AbstractContext subclass or None
        """
        if not isinstance(context_type, ContextType):
            raise ValueError('context_type must be an instance of ContextType')

        if context_type in _ctx.store and len(_ctx.store[context_type]) > 0:
            return _ctx.store[context_type][-1]

    @staticmethod
    def get_or_default(context_type, default):
        """Returns most recent instance of given Context Type or default

        :param context_type: Context Type to retrieve from the store
        :param default: value to return in case given context does not exist
        :return: an instance of AbstractContext subclass or default
        """
        ctx = ContextStore.get_context(context_type)
        return ctx if ctx else default

    @staticmethod
    def get_or_empty(context_type):
        """Returns recent instance of given Context Type or an empty context

        :param context_type: Context Type to retrieve from the store
        :return: an instance of AbstractContext subclass
        """
        return ContextStore.get_or_default(context_type, _emptyContext())

    @staticmethod
    def _add_context(ctx):
        """Add given instance of AbstractContext subclass to context store

        This private member is only used by ``AbstractContext``

        :param ctx:
        :return: None
        """
        if not isinstance(ctx, AbstractContext):
            raise ValueError('ctx must be be an instance of AbstractContext')

        context_type = ctx.get_type()
        if context_type not in _ctx.store:
            _ctx.store[context_type] = collections.deque()

        _ctx.store[context_type].append(ctx)

    @staticmethod
    def _remove_context(ctx):
        """Removes the most recently added context of given Context Type

        This private member is only used by ``AbstractContext``
        :param ctx:
        :return: None
        """
        if not isinstance(ctx, AbstractContext):
            raise ValueError('ctx must be be an instance of AbstractContext')

        context_type = ctx.get_type()

        if context_type not in _ctx.store or len(_ctx.store[context_type]) == 0:
            return

        _ctx.store[context_type].pop()

```

Does it make sense?

Reply:
Looks cool. Would love to see it in action :)